### ISSUE TITLE
Callback TerminateOnNan

### ISSUE THREAD

**Issue by taleslimaf** [skorch-dev/skorch#627]
I'm using Skorch and in a few moments my model starts showing NaN in the Parameters.
He trains very well and when he is about to converge on a solution he starts presenting NaN.
By lowering the lr I get better results, which do not give NaN, but as a consequence I have to increase the number of epochs.

I would like to know if it would be possible to do something similar to the link:
https://github.com/pytorch/ignite/blob/master/ignite/handlers/terminate_on_nan.py

When the model shows NaN, I get to its previous state, where it still showed different results from NaN.

**Comment by BenjaminBossan** [comment]
Here is a proposal:

```python
from skorch.callbacks import Callback, Checkpoint

class TerminateOnNan(Callback):
    """Interrupts training when nans or infs are encountered"""
    def on_batch_end(self, net, *args, y_pred, loss, **kwargs):
        if not (torch.isfinite(loss) and torch.isfinite(y_pred).all()):
            raise KeyboardInterrupt

cp = Checkpoint(dirname='checkpoints')

net = NeuralNe(
    ...
    callbacks=[cp, TerminateOnNan()],
)

net.fit(X_regr, y_regr)  # stops on nan or inf
net.load_params(checkpoint=cp)
```

Would that work for you?

In general, I have a bad feeling when I have a net that is close to creating nans or infs. Maybe you should check if some form of regularization could help prevent those, like weight decay or [gradient clipping](https://skorch.readthedocs.io/en/stable/callbacks.html#skorch.callbacks.GradientNormClipping).

**Comment by taleslimaf** [comment]
I am working on my PhD with a peculiar activation function. :D 
It is converging for very good results, but in some scenarios, even with low lr, if the number of epochs is too much, it converges the weights to NaN.

I am still doing the research, so I would like to pick up the network before it starts giving these problems to analyze what I should do.

I converged to this solution here:

```python
from skorch.callbacks import Callback

class TerminateOnNan(Callback):

    def __init__(self, sink=print):
        self.sink = sink
        
    # pylint: disable=arguments-differ
    def on_grad_computed(self, net, named_parameters, **kwargs):
        parameters = (p for _, p in named_parameters)       
        
        if isinstance(parameters, torch.Tensor):
            parameters = [parameters]
            
        parameters = list(filter(lambda p: p.grad is not None, parameters))
        values = np.stack([p.grad.detach().abs().max().numpy() for p in parameters], axis=0)
        
        if self._is_nan_or_inf(values):
            if net.verbose:
                self._sink("Output contains NaN or Inf. Stop training.",
                           verbose=net.verbose)
            raise KeyboardInterrupt
        
    def _is_nan_or_inf(self, values):
        return np.isnan(values).any() or np.isinf(values).any()
    
    def _sink(self, text, verbose):
        #  We do not want to be affected by verbosity if sink is not print
        if (self.sink is not print) or verbose:
            self.sink(text)   
```

Your code is very interesting and helped me get to this one of mine. The only problem with yours is that it stops the network already with NaN values. What I needed was a version that caught the network before it converged on NaN. Thanks a lot for the help. I believe that, if it is useful for other people, this Callback could be in the package.

**Comment by BenjaminBossan** [comment]
Glad you found a solution that worked for you.

I was thinking to go through the gradients as you did but then I thought that even if all grads are finite, it could still happen that the weights or model outputs are not (at least I think it could). So I went with my solution, which is a bit more clunky but might catch a few cases that yours doesn't.

> `p.grad.detach().abs().max().numpy()`

Why do you need all of that? Wouldn't something like `all(torch.isfinite(grad).all() for grad in grads)` work or is your solution faster?

> I believe that, if it is useful for other people, this Callback could be in the package.

I think it would need some polishing (maybe also combine it with my callback to cover more cases) but could be useful. Tests are not super trivial for this one. If you want to, you could work on that.

**Comment by taleslimaf** [comment]
Yes, this has happened in some cases here. Even stopping training before the grad explodes.
Despite this, they are very rare cases, which is already easier for me to make decisions about the activation function.

The inspiration for my code came from here:
https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py

But I believe that your line of code is computationally faster. Smaller number of operations. I'll make the switch, as it makes a lot more sense.

Thank you very much for your help.

**Comment by supetronix** [comment]
@BenjaminBossan 
Is there a plan to make @taleslimaf suggested solution, with your optimization, part of Skorch?

**Comment by BenjaminBossan** [comment]
@supetronix It's not currently planned. If you want to work on it, feel free to create a PR. We're ready to help if any question should arise.
